crates/typst-utils/src/format.rs

```diff
@@ -0,0 +1,139 @@
+use std::{cmp::min, time::Duration};
+
+/// Returns number of `(days, hours, minutes, seconds, milliseconds, microseconds)`.
+fn get_duration_parts(duration: &Duration) -> (u16, u8, u8, u8, u16, u16) {
+    // In practice we probably don't need nanoseconds.
+    let micros = duration.as_micros();
+    let (millis, micros) = (micros / 1000, (micros % 1000) as u16);
+    let (sec, millis) = (millis / 1000, (millis % 1000) as u16);
+    let (mins, sec) = (sec / 60, (sec % 60) as u8);
+    let (hours, mins) = (mins / 60, (mins % 60) as u8);
+    let (days, hours) = ((hours / 24) as u16, (hours % 24) as u8);
+    (days, hours, mins, sec, millis, micros)
+}
+
+/// Format string using `days`, `hours`, `minutes`, `seconds`.
+fn format_dhms(days: u16, hours: u8, minutes: u8, seconds: u8) -> String {
+    match (days, hours, minutes, seconds) {
+        (0, 0, 0, s) => format!("{s:2} s"),
+        (0, 0, m, s) => format!("{m:2} m {s:2} s"),
+        (0, h, m, s) => format!("{h:2} h {m:2} m {s:2} s"),
+        (d, h, m, s) => format!("{d:3} d {h:2} h {m:2} m {s:2} s"),
+    }
+}
+
+/// Format string starting with number of days and going bigger from there.
+fn time_starting_with_seconds(duration: &Duration) -> String {
+    let (days, hours, minutes, seconds, _, _) = get_duration_parts(duration);
+    format_dhms(days, hours, minutes, seconds)
+}
+
+/// Format string starting with number of milliseconds and going bigger from there.
+/// `precision` is how many digits of microseconds from left will be preserved.
+/// Note that this function will always remove all trailing zeros for microseconds.
+fn time_starting_with_ms_with_precision(duration: &Duration, precision: usize) -> String {
+    let digits = precision;
+    let (d, h, m, s, ms, mcs) = get_duration_parts(duration);
+    let mcs_digits = {
+        let mut mcs = mcs;
+        println!("before {mcs}");
+        while mcs % 10 == 0 && mcs != 0 {
+            mcs /= 10
+        }
+        println!("after {mcs}");
+        mcs.to_string().len()
+    };
+    let ms = format!(
+        "{1:.0$} ms",
+        if mcs != 0 { min(digits, mcs_digits) } else { 0 },
+        ms as f32 + mcs as f32 / 1000.0
+    );
+    match (d, h, m, s) {
+        (0, 0, 0, 0) => ms.to_string(),
+        (d, h, m, s) => format_dhms(d, h, m, s),
+    }
+}
+
+fn time_as_ms(duration: &Duration) -> String {
+    time_starting_with_ms_with_precision(duration, 2)
+}
+
+pub fn elapsed_time(duration: &Duration) -> String {
+    time_starting_with_seconds(duration)
+}
+
+pub fn eta_time(duration: &Duration) -> String {
+    time_starting_with_seconds(duration)
+}
+
+pub fn compilation_time(duration: &Duration) -> String {
+    time_as_ms(duration)
+}
```

Comment:
I believe it should be fine to inline these functions into their respective usages. Makes more sense for the format choice to be specified where they're used instead of in a different module.

Reply:
I know that it's "fine" to do that, I'm just thinking that it's easier to read and understand `eta = elapsed_time(elapsed_duration)` rather than using `time_starting_with_seconds`.

If you are sure that it won't decrease readability or increase maintainability or something, then I can remove these wrappers. I just don't think that they are that bad, but also just a function name alias isn't something that I see everyday.

Reply:
> I know that it's "fine" to do that, I'm just thinking that it's easier to read and understand `eta = elapsed_time(elapsed_duration)` rather than using `time_starting_with_seconds`.

It could make sense if the functions were used more than once, but they're used only once, and in that sense it's not helpful as the name doesn't explain what's going on. It's something like `let thing = thing()` - it doesn't say much, just that the code for it is elsewhere.

Reply:
Fine, I'll get rid of them.

Reply:
Just so you know, `compilation_time` is used twice and therefore inlining the base function leads to the usage of precision in 2 very close places. So I think that leaving either `time_starting_with_ms` or `compilation_time` wrapper for the function with a precision parameter isn't that useless already.

Reply:
I'm ok with having a single public wrapper. Just multiple public wrappers with the ~same stuff is a bit too much for me.

Reply:
Eh, I already changed to accepting it, since it used precision twice before: https://github.com/typst/typst/pull/4587/files#diff-ea95583ebe87e25ee323e9ad5d1a654f2fbf243e9f4d1727c49d7271585149bdL300-R310. If we get more than these 2 usages, then it is probably wise to add a wrapper. If they all use the same precition.